test/rulesets/Base/indexing.jl

```diff
@@ -77,6 +77,91 @@
+        xfix = [Base.Fix1(*, pi), Base.Fix1(^, ℯ), Base.Fix1(/, -1)]
+        dxfix = [Tangent{Base.Fix1}(; x = i/10) for i in 1:3]
+        # test_frule(getindex, xfix ⊢ dxfix, 1)
+        # test_rrule(getindex, xfix ⊢ dxfix, 1)
+        
+        dx1 = unthunk(rrule(getindex, xfix, 1)[2](dxfix[1])[2])
+        @test dx1[1] == dxfix[1]
+        @test iszero(dx1[2])
+        
+        dx23 = unthunk(rrule(getindex, xfix, 2:3)[2](dxfix[2:3])[2])
+        @test iszero(dx23[1])
+        @test dx23[3] == dxfix[3]
+    end
+
+    @testset "second derivatives: ∇getindex" begin
+        @eval using ChainRules: ∇getindex
+        # Forward, scalar result
+        test_frule(∇getindex, rand(2, 3), rand(), 3)
+        test_frule(∇getindex, rand(2, 3), rand()+im, 2, 1)
+        # array result
+        test_frule(∇getindex, rand(2, 3), rand(2), 4:5)
+        test_frule(∇getindex, rand(2, 3), rand(3), 1, :)
+        test_frule(∇getindex, rand(2, 3), rand(1, 2), [CartesianIndex(2, 1) CartesianIndex(2, 2)]  ⊢ NoTangent())
+        test_frule(∇getindex, rand(2, 3), rand(3), Bool[1 0 1; 0 1 0])
+        # arrays of arrays
+        test_frule(∇getindex, [rand(2) for _ in 1:3], rand(2), 3)
+        test_frule(∇getindex, [rand(2) for _ in 1:3], [rand(2), rand(2)], 1:2)
+
+        # Reverse, scalar result
+        test_rrule(∇getindex, rand(2, 3), rand(), 3; check_inferred=false)
+        test_rrule(∇getindex, rand(2, 3), rand()+im, 2, 1; check_inferred=false)
+        # array result
+        test_rrule(∇getindex, rand(2, 3), rand(2), 4:5; check_inferred=false)
+        test_rrule(∇getindex, rand(2, 3), rand(3), 1, :; check_inferred=false)
+        test_rrule(∇getindex, rand(2, 3), rand(1, 2), [CartesianIndex(2, 1) CartesianIndex(2, 2)]  ⊢ NoTangent(); check_inferred=false)
+        test_rrule(∇getindex, rand(2, 3), rand(3), Bool[1 0 1; 0 1 0]; check_inferred=false)
+        # arrays of arrays
+        test_rrule(∇getindex, [rand(2) for _ in 1:3], rand(2), 3; check_inferred=false)
+        test_rrule(∇getindex, [rand(2) for _ in 1:3], [rand(2), rand(2)], 1:2; check_inferred=false)
+    end
+
+    @testset "GPU" begin
+        x_23_gpu = jl(rand(2, 3))
+    
+        # Scalar indexing, copied from:  @macroexpand @allowscalar A[i]
+        # Gives an error in Pkg.test, no idea why
+        # y1, bk1 = rrule(CFG, Base.task_local_storage, () -> x_23_gpu[1], :ScalarIndexing, ScalarAllowed)
+        # @test y1 == @allowscalar x_gpu[1]
+        # bk1(1.0)  # This is zero, because finite-differencing ignores the function
+        # ... but this works, and calls the rule:
+        # Zygote.gradient(x -> @allowscalar(x[1]), jl(rand(3)))[1]
+        
+        y2, bk2 = rrule(getindex, x_23_gpu, :, 2:3)  # fast path, just broadcast .+=
+        @test unthunk(bk2(jl(ones(2,2)))[2]) == jl([0 1 1; 0 1 1])
+
+        y3, bk3 = rrule(getindex, x_23_gpu, 1, [1,1,2])  # slow path, copy to CPU
+        @test_skip Array(y3) == Array(x_gpu)[1, [1,1,2]]  # error in Pkg.test, no idea why
```

Comment:
Can we reproduce and open an issue on Julia itself and link back here?

Reply:
After fiddling a bit, here's a better version. These steps work in global scope, but fail inside the let block. FiniteDifferences is involved, but not sure where the blame is:
```
julia> let
               x_23_gpu = jl(rand(2, 3))  # using JLArrays, loaded for @gpu in test_helpers.jl
           
               # Scalar indexing, copied from:  @macroexpand @allowscalar A[i]
               y1, bk1 = rrule(CFG, Base.task_local_storage, () -> x_23_gpu[1], :ScalarIndexing, ScalarAllowed)
                @test y1 == @allowscalar x_23_gpu[1]
                bk1(1.0)
       end
ERROR: StackOverflowError:
Stacktrace:
     [1] to_vec(x::JLArray{Float64, 2})
       @ FiniteDifferences ~/.julia/packages/FiniteDifferences/VpgIT/src/to_vec.jl:73
     [2] to_vec(x::Base.ReshapedArray{Float64, 1, JLArray{Float64, 2}, Tuple{}})
       @ FiniteDifferences ~/.julia/packages/FiniteDifferences/VpgIT/src/to_vec.jl:84
--- the last 2 lines are repeated 39990 more times ---
 [79983] to_vec(x::JLArray{Float64, 2})
       @ FiniteDifferences ~/.julia/packages/FiniteDifferences/VpgIT/src/to_vec.jl:73
```